### PR TITLE
[rllib] Remove need to pass around registry

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Avoid the need to pass the dict of registered objects around by fetching them from global redis.

### CODE REVIEW

**Comment by robertnishihara** [comment]
I'm very skeptical of this change. RLlib should stick to using Ray APIs. Relying on implementation details like this will make it hard to change the implementation later on.

Also, using out-of-band mechanisms like this will compromise properties like fault tolerance.

**Comment by ericl** [comment]
We're not tied to any particular implementation here -- the requirement is that we support a _global_ registry. Right now it uses redis, like named actors, but we can easily switch it to use the object store or some other mechanism in the future.

**Comment by ericl** [comment]
I updated this to refactor both this and named actors to use an internal KV interface.

**Comment by skyler0712** [comment]
Are you guys ever going to get done fixing my account

Skyler Bays

**Comment by robertnishihara** [comment]
@ericl that seems like a good approach for now.

Can you say why you introduced the `post_init_hooks`? What is that needed for and how were you solving that problem before?

**Comment by ericl** [comment]
The motivation for the hook is that we need register calls to work before
ray has initialized (most generally, on import or modules). Hence, we need
to queue puts and flush them post Ray init.

**Review by robertnishihara — APPROVED** [review]
Looks fine to me assuming tests pass.

File: python/ray/tune/registry.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 from types import FunctionType
 
+import cloudpickle
```
Comment (robertnishihara) [inline python/ray/tune/registry.py]: `cloudpickle` technically isn't a dependency of Ray (e.g., it is not in our `setup.py`), so users may not have it installed.

We have our own cloudpickle which you can import with e.g., `import ray.cloudpickle as pickle`. The point of doing this is that it's very easy to have a cloudpickle version mismatch on different machines, but if we ship our own then we don't run into this problem.

Comment (ericl) [reply python/ray/tune/registry.py]: Fixed!

**Comment by richardliaw** [comment]
Lint and a couple tests are failing on Travis...

**Pull request merged**